### PR TITLE
Cross-contract calling: simple debugger

### PR DESCRIPTION
**Pull request by pmikolajczyk41**
# Description

## Motivation

While full, true contract debugging is currently not achievable, we still can try to get some insights from an outside client to the contracts pallet's internals. Since it is the pallet that owns and maintains call stack, we can take an advantage of this direct access. 

### Target environment/client

Surely, when we work in a full-stack setting (i.e. including node layer), there's no point in introducing breakpoints. In the case of an RPC call, we could try returning some trace, similarly to a debug buffer. However, my intention is to achieve more interactive experience.

Suppose that we own/hold the whole runtime (embedded within externalities) and have direct (synchronous) access to it. Then, we can establish a simple callback-based communication with the contracts pallet. This is exactly the case when working with tools like https://github.com/Cardinal-Cryptography/drink, but also for pallet's unit tests, or even `try-runtime` suite.

### Workflow

For now, we add just two simple breakpoints: just before and after an `Executable` object is invoked. Hence, we can debug complex multi-contract execution traces. Since the whole runtime is synchronous, naturally the callbacks will stop execution and progress when the handling is done.

## Plan

In the near future we plan to add more such breakpoints, providing further information, but firstly we want to check the approach and define what things would be useful.

# Checklist

- [x] My PR includes a detailed description as outlined in the "Description" section above
- [ ] My PR follows the [labeling requirements](https://github.com/paritytech/substrate/blob/master/docs/CONTRIBUTING.adoc#merge-process) of this project (at minimum one label for each `A`, `B`, `C` and `D` required)
- [x] I have made corresponding changes to the documentation (if applicable)
- [ ] I have added tests that prove my fix is effective or that my feature works (if applicable)
- [ ] If this PR alters any external APIs or interfaces used by Polkadot, the corresponding Polkadot PR is ready as well as the corresponding Cumulus PR (optional)

### CODE REVIEW

**Comment by athei** [comment]
I think this is a good idea. One thing that is important to me is that all this debugging code doesn't increase the complexity of the production code too much. What I mean by that is that we don't end up with a lot of compile time features and associated types added to `Config`.

I expect that over time you come up with more things you want to add to this debugging interface. Therefore I think we should make sure that we always only have a single feature and associated type.

Concretely:

1. Rename feature to `unsafe-debug` so that it will enable/disable all debugging (except debug buffer which is OK for production use).

2. Make sure all the debugging traits are grouped by a single trait to we always only need a single associated type on the `Config`:

```rust
trait Debug: ExecutionObserver + ExecutionTrace + ... 
{
}
```

The downside is that all traits need to be implemented but we can just have sensible default implementations.

**Comment by pmikolajczyk41** [comment]
In order to provide some reasonable usecase, I prepared an example scenario of cross contract calling and debugged it using drink library (after necessary on-spot enrichments). The code is available on a branch: https://github.com/Cardinal-Cryptography/drink/tree/pallet-breakpoint.

In the example we have 3 simple contacts: `a`, `b` and `c`. In the scenario, we call contract `a` with some `u32` value. Contract a forwards the argument to `b`, `b` forwards it to `c` and `c` returns a result of some arithmetic operation, which is then returned up the stack to the caller.

_Note: since it was faster to use call builder pattern in the contracts, I also passed the addresses of `b` and `c` along the `u32` argument (so that delegation is possible)._

After running `./debugger_showcase.sh` you should see:
```
Building contract c
Building contract b
Building contract a
Drinking

running 1 test

successes:

---- tests::test stdout ----
Contract with hash `0x118a…8b5e` has been called with data: 
    new
and returned:
    ()

Contract with hash `0x2b1e…9b14` has been called with data: 
    new
and returned:
    ()

Contract with hash `0x38fa…c6bc` has been called with data: 
    new
and returned:
    ()

Contract with hash `0x118a…8b5e` has been called with data: 
    call { arg: 7 }
and returned:
    Ok(22)

Contract with hash `0x2b1e…9b14` has been called with data: 
    call { account_c: 5HHtX3hsCmVUTzpQtQKTnNyAQLnVzgUNrefA1kYvFwPfYbMX, arg: 7 }
and returned:
    Ok(22)

Contract with hash `0x38fa…c6bc` has been called with data: 
    call { account_b: 5CjYz6jzGja6WnwTWgC11mMCSyscknKoAb3JpmuCQ3TfUs3m, account_c: 5HHtX3hsCmVUTzpQtQKTnNyAQLnVzgUNrefA1kYvFwPfYbMX, arg: 7 }
and returned:
    Ok(22)
```

The hardest part is to enable arbitrary handler. Why the hardest? Because the handler is invoked by the runtime. And since it is passed as an associated type to the pallet configuration, it must be fixed - we cannot have e.g. generic runtime. The way I chose to solve the problem (and actually there's the only one I know) is to outsource handler via runtime interface to a specially registered externalities extension. That way we can pass to the drink's `Session` any type that implements certain trait.

File: frame/contracts/src/exec.rs
Comment (athei) [inline frame/contracts/src/exec.rs:911]: Only clone when compiled with debugging. We can also think about not cloning the input at all and leave that to the debugger to store the input if needed. We already pass it in `before_call` and it doesn't change in between.

```suggestion
			#[cfg(feature = "unsafe-debug")]
			let (code_hash, input_clone) = {
			    let code_hash = executable.code_hash().clone();
			    T::Debug::before_call(&code_hash, entry_point, &input_data);
			    (code_hash, input_data.clone())
			};
```

Comment (pgherveou) [inline frame/contracts/src/exec.rs:911]: thought, maybe to keep things dry we could do something like this in pseudo code
```
let debug_call = T::Debug::start_call(...)
...
debug_call.end(&output)
```

File: frame/contracts/src/exec.rs
Comment (athei) [inline frame/contracts/src/exec.rs:919]: We can move the input vector since we don't need it anymore. In case the debugger wants to store the data.

```suggestion
			T::Debug::after_call(&code_hash, entry_point, input_clone, &output);
```

File: frame/contracts/src/exec.rs
Comment (athei) [inline frame/contracts/src/exec.rs:915]:
```suggestion
				.execute(self, &entry_point, input_data)
```

File: frame/contracts/src/lib.rs
Comment (pgherveou) [inline frame/contracts/src/lib.rs:142]: this is not needed right?

File: frame/contracts/src/exec.rs
Comment (pmikolajczyk41) [reply frame/contracts/src/exec.rs]: :white_check_mark:

File: frame/contracts/src/exec.rs
Comment (pmikolajczyk41) [reply frame/contracts/src/exec.rs]: :white_check_mark:

File: frame/contracts/src/exec.rs
Comment (pmikolajczyk41) [reply frame/contracts/src/exec.rs]: I applied the code suggestion. I wouldn't rely on the debugger to store input, since it might make sense to actually implement only `after_call` or have stateless debugger

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3311091

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3311179

**Comment by pmikolajczyk41** [comment]
now I'm getting:
```
error[E0437]: type `Debug` is not a member of trait `pallet_contracts::Config`
    --> bin/node/runtime/src/lib.rs:1262:2
     |
1262 |     type Debug = ();
     |     ^^^^^^^^^^^^^^^^ not a member of trait `pallet_contracts::Config`
```

after adding the type declaration (previously it was complaining about **not** including this member). I guess that this kitchen-sink thing is compiled in multiple configurations, but I don't know how to get it right :(

at least test passes correctly (at least locally for me :stuck_out_tongue: )

File: frame/contracts/src/exec.rs
Comment (pgherveou) [reply frame/contracts/src/exec.rs]: > this debug_call variable would have to be feature-gated as well, right?

yes so

```
#[cfg(feature = "unsafe-debug")]
let debug_call = T::Debug::start_call(...)
...
#[cfg(feature = "unsafe-debug")]
debug_call.end(&output)
```

to me more correct

Comment (pmikolajczyk41) [reply frame/contracts/src/exec.rs]: the problem I see with this approach is that we put more requirements for the user (and mock implementations) to implement another type for this `debug_call`

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3313848

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3315822

File: frame/contracts/Cargo.toml
Comment (athei) [inline frame/contracts/Cargo.toml:58]: No. Just conditionally compile the tests.

**Review by athei — CHANGES_REQUESTED** [review]

File: frame/contracts/Cargo.toml
Comment (pgherveou) [reply frame/contracts/Cargo.toml:58]: so this will never run in CI? which makes the test less useful..

Comment (athei) [reply frame/contracts/Cargo.toml:58]: Doesn't the CI run `cargo test --all-features`?

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3319296

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3319295

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3319386

File: frame/contracts/Cargo.toml
Comment (pgherveou) [reply frame/contracts/Cargo.toml]: 🤔  I thought not.
Some other folks are doing similar things 
https://github.com/paritytech/polkadot/pull/7538/files/b72c655f9323004d8da523f6f4690204172e9b83#r1275964591

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3323056

File: frame/contracts/Cargo.toml
Comment (athei) [reply frame/contracts/Cargo.toml]: The proper solution is to change the CI script to enable all features. This just doesn't seem like intended behaviour.

Comment (pmikolajczyk41) [reply frame/contracts/Cargo.toml]: I'm not sure where how I should change it then :|

Comment (pmikolajczyk41) [reply frame/contracts/Cargo.toml]: added; the corresponing CI job succeeded, but I have no proof in logs for running this particular testcase - the log is truncated

but to be 100% I will corrupt an assertion in the testcase and see whether it is reported

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3324722

File: frame/contracts/Cargo.toml
Comment (pmikolajczyk41) [reply frame/contracts/Cargo.toml]: confirmed: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3324722

reverting change then; @athei @pgherveou so I guess we are done here

File: frame/contracts/src/exec.rs
Comment (athei) [inline frame/contracts/src/exec.rs:921]: Let's move in the input as we don't need it anymore. In case the observer wants to store it there is no clone necessary.

```suggestion
			T::Debug::after_call(&code_hash, entry_point, input_clone, &output);
```

File: frame/contracts/src/lib.rs
Comment (athei) [reply frame/contracts/src/lib.rs:142]: Re-export from the `unsafe_debug` module then.

File: frame/contracts/src/lib.rs
Comment (athei) [inline frame/contracts/src/lib.rs:100]: Move it to the module please. Just a style nit but looks nicer in my opinion.

File: frame/contracts/src/lib.rs
Comment (athei) [inline frame/contracts/src/lib.rs:352]: Docs needed.

File: bin/node/runtime/Cargo.toml
Comment (athei) [inline bin/node/runtime/Cargo.toml:71]: Is there a reason why we set this here? Isn't it covered by tests + CI anyways?

Comment (pmikolajczyk41) [reply bin/node/runtime/Cargo.toml:71]: it's the only configuration that allowed me to pass both commands (which are triggered in CI):
 - `SKIP_WASM_BUILD=1 cargo check --locked --benches --all -p kitchenskin-runtime`
 - `cargo check --locked --benches --all -p kitchenskin-runtime`

File: frame/contracts/src/exec.rs
Comment (pmikolajczyk41) [reply frame/contracts/src/exec.rs]: changed; however, this made `before_call` and `after_call` arguments differ, but I think it might make sense to move input after call

File: frame/contracts/src/lib.rs
Comment (pmikolajczyk41) [reply frame/contracts/src/lib.rs]: moved

File: frame/contracts/src/lib.rs
Comment (pmikolajczyk41) [reply frame/contracts/src/lib.rs:356]: added

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3326497

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3326483

**Comment by pmikolajczyk41** [comment]
bot rebase

File: bin/node/runtime/Cargo.toml
Comment (athei) [reply bin/node/runtime/Cargo.toml:71]: That would be very had as it means the pallet requires that feature. For me it works without that feature. You need to remove `type Debug = ()` of course.

**Review by athei — COMMENTED** [review]
Can you move the tests for the debugging feature to its own sub module of tests? Then we don't have to sprinkle the conditional compilation that much. Basically having:

`tests.rs`
`tests/unsafe_debug.rs`

File: bin/node/runtime/Cargo.toml
Comment (pmikolajczyk41) [reply bin/node/runtime/Cargo.toml]: it's failing though: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3331887

that is really pain in the back and I'm struggling with it from the very beginning; no idea why it requires this `Debug` type here

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3331877

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3331879

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3331878

File: bin/node/runtime/Cargo.toml
Comment (athei) [reply bin/node/runtime/Cargo.toml]: Ahh yeah of course because we manually enable the feature there for this crate. Yeah you have to roll back this change :)

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3332279

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3332277

File: bin/node/runtime/Cargo.toml
Comment (pmikolajczyk41) [reply bin/node/runtime/Cargo.toml]: @athei while the two commands are now passing for me as well, for some reason, rust doc is requiring this item - you can run:
```
cargo doc --locked --workspace --all-features --no-deps
```

Nothing like:
```rust
#[cfg(doc)]
type Debug = ();
// or
#[cfg(feature = "unsafe-debug")]
type Debug = ();
```
helps

Comment (athei) [reply bin/node/runtime/Cargo.toml]: You didn't roll back the change yet.

Comment (pmikolajczyk41) [reply bin/node/runtime/Cargo.toml]: I think I did - there's no change to `/bin/node` directory

File: frame/contracts/src/exec.rs
Comment (pgherveou) [inline frame/contracts/src/exec.rs:349]: are these new trait required?

**Review by pgherveou — APPROVED** [review]

**Review by athei — APPROVED** [review]

**Comment by athei** [comment]
We need approval from CI team then we can merge.

**Review by mateo-moon — APPROVED** [review]

**Review by rcny — APPROVED** [review]

**Comment by pgherveou** [comment]
bot merge

**Pull request merged**
